BrainPortal/app/models/boutiques_cluster_task.rb

```diff
@@ -0,0 +1,354 @@
+      userfile_id = invoke_params[input.id]
+      next if userfile_id.blank? # that happens when it's an optional file
+      userfile    = Userfile.find(userfile_id)
+
+      # Most common situation
+      if ! input.list || ! userfile.is_a?(CbrainFileList)
+        make_available(userfile, userfile.name)
+        next
+      end
+
+      # In case the input is a list and is assigned a CbrainFileList
+      userfile.sync_to_cache
+      userfile_list = cbrainfilelist.userfiles_accessible_by_user!(user, nil, nil, file_access_symbol)
+      userfile_list.compact.each do |subfile|
+        make_available(subfile, subfile.name)
+      end
+    end
+
+    true
+  end
+
+  def cluster_commands
+    descriptor = self.descriptor_for_cluster_commands
+
+    invoke_struct = self.invoke_params.dup
+
+    # Replace userfile IDs for file basenames in the invoke struct
+    descriptor.file_inputs.each do |input|
+      userfile_id = invoke_params[input.id]
+      next if userfile_id.blank? # that happens when it's an optional file
+      userfile    = Userfile.find(userfile_id)
+
+      # Most common situation
+      if ! input.list || ! userfile.is_a?(CbrainFileList)
+        invoke_struct[input.id] = (input.list ? [ userfile.name ] : userfile.name)
+        next
+      end
+
+      # In case the input is a list and is assigned a CbrainFileList
+      userfile.sync_to_cache
+      userfile_list = cbrainfilelist.userfiles_accessible_by_user!(user, nil, nil, file_access_symbol)
+      subnames = userfile_list.compact.map(&:name)  # [ 'userfilename1', 'userfilename2' ... ]
+      invoke_struct[input.id] = subnames
+    end
+
+    # Replace the "0"/"1" strings we use for booleans with true and false
+    descriptor.inputs.select { |input| input.type == 'Flag' }.each do |input|
+      next if invoke_struct[input.id].blank?
+      invoke_struct[input.id] = true  if invoke_struct[input.id] == '1'
+      invoke_struct[input.id] = false if invoke_struct[input.id] == '0'
+    end
+
+    # Write down the file with the invoke struct
+    invoke_json_basename = "invoke.#{self.run_id}.json"
+    File.open(invoke_json_basename ,"w") do |fh|
+      fh.write JSON.pretty_generate(invoke_struct)
+      fh.write "\n"
+    end
+
+    boutiques_json_basename = "boutiques.#{self.run_id}.json"
+    File.open(boutiques_json_basename, "w") do |fh|
+      fh.write JSON.pretty_generate(descriptor)
+      fh.write "\n"
+    end
+
+    if self.boutiques_bosh_exec_mode == :simulate # the default
+      simulate_com = <<-SIMULATE
+        bosh exec simulate
+          -i #{invoke_json_basename.bash_escape}
+          #{boutiques_json_basename.bash_escape}
+      SIMULATE
+      simulate_com.gsub!("\n"," ")
+      simulout = IO.popen(simulate_com) { |fh| fh.read }
+      simulout.sub(/^Generated.*\n/,"") # header junk from simulate
+      commands = <<-COMMANDS
+        # Main tool command, generated with bosh exec simulate
+        #{simulout.trim}
```

Comment:
I have undefined method trim errors

Reply:
It's a bug I will fix right now

Reply:
Mea culpa. That will teach me about making last minute changes.